htlcswitch/link.go

```diff
@@ -627,7 +627,8 @@ func (l *channelLink) EligibleToForward() bool {
 func (l *channelLink) EligibleToUpdate() bool {
 	return l.channel.RemoteNextRevocation() != nil &&
 		l.ShortChanID() != hop.Source &&
-		l.isReestablished()
+		l.isReestablished() &&
+		l.quiescer.canSendUpdates()
```

Comment:
I don't think atomic bools are the answer but I do think that you're right that I need to reexamine this with considerations for races.

Reply:
Examining this more in depth the main issue regarding atomicity is that each one of these function calls is read independently. Replacing it with atomic bools will not fix this.

This reminds me that we need to be much more consistent about how we do locking and ideally rid ourselves of locking entirely for things that have their own live threads like the channel link. I am going to add an RLock to this method which should solve the problem just fine since multiple RLocks can be held. This will prevent self-deadlocking behavior when composing APIs out of one another. It should be noted however that this does not work for a composition of writes since afaiu the golang runtime does not safely allow the same thread to lock the same mutex multiple times.

Assuming we stick with logs rather than channel based synchronization using event loops, The correct way to handle this in a broader sense is to duplicate all methods with private and public variants, where the public variant is responsible for acquiring the type of lock needed and the inner ones are all unchecked wrt locking behavior. However, this can add lots of overhead and I am not ready to conduct an architectural overhaul on the link just yet. I'm still working on doing it for `LightningChannel`

Reply:
Nvm. Recursive read locking doesn't work either... due to fairness guarantees by RWMutex. GG.

Reply:
I have fixed the underlying issue to a first approximation in #8990. This PR now depends on that but it will go in separately because this is an issue that actually precedes this PR.

Reply:
I've now decided to handle this using the "closure throwing" approach.